blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+instructions are required (if the bug is something non-obvious like
+unexpected behavior, some comments on what the expected behavior was may
+still be helpful), since the `rr` trace is guaranteed to perfectly capture
+the environment the bug was reproduced in. This almost immediately knocks
+out all the common problems I started this post with. "Works for me" is no
+longer an available answer. If it's in the trace, it broke on somebody's machine
+and can be debugged. "Heisenbug"s are no longer an issue. If it was captured
+once in `rr`, it can always be debugged. It even solves the busy-expert problem,
+as it allows non-experts to help out with triage. If such a report does not
+contain an `rr` trace, any developer, and in particular non-experts can attempt to
+reproduce the bug and create one. Even if the expert is still required to do
+the final diagnosis, doing so from an `rr` trace is orders of magnitude faster
+than from a plain bug report.
+
+# Chronomancy for dummies
+
+A computer is fundamentally a deterministic machine. Given equivalent
+states as inputs, most instructions will produce a deterministic state
+as output. So where do all the subtle execution differences come from that
+produce diverging executions and prevent bugs from reproducing? Well, the
+complete answer is complicated and has lots of details, but is roughly split
+up into:
+
+1) The input state is large. Probably the state of your entire hard drive and
+    memory. That's at least `2^2^large` bits of state right there each of which
+    could potentially cause a difference in execution. You really want to figure
+    out what state is relevant, since `2^2^large` is probably too many bits to send
+    somebody.
+
+2) Any input or other asynchronous events. This can mean user input, or
+    data from the network (via the NICs) or other devices. It also means things like
+    the timing of interrupts.
+
+3) Execution ordering and data races in multi-threaded executions.
+
+However, in theory if a tool was able to capture 100% of the relevant state
+from these categories, it could repeated generate exactly the same memory image.
+This is not a novel idea, but the devil is in the detail. As already mentioned,
+just capturing everything is probably too large, if at all possible (e.g. how
+do you capture interrupt timing - how do you even talk about time in this setting?).
+
+In other words, for some input state $\mathcal{S_{i}}$,
+the output state $\mathcal{S_{i+1}}$ is determined by some pure, deterministic
+function $\mathcal{I}_{ip(\mathcal{S}_i)}$, corresponding to the instruction to be
+executed at time $i$, or some asynchronous event $\mathcal{A}_i$
+
+$$ \mathcal{S}_{i+1} = \begin{cases}
+    \mathcal{A}_i(S_{i}) & \text{if an asynchronous event occured at time $i$} \\
+    \mathcal{I}_{\text{ip}(\mathcal{S}_i)} & \text{otherwise}
+\end{cases} $$
+
+Did that help? No? oh. Well, it made me feel better about my physics degree and
```

Comment:
Possible alternative wording, perhaps:
```suggestion
Assuming we can discretize the asynchronous event into..... Okay, I realize this probably isn't helping. But I've been told it's not research until it has at least one equation in it and it made me feel better about my physics degree. Also, people worked really hard to get that nice equation rendering working, so I better use it. Anyway, where were we?
```

Reply:
Doesn't quite sound like me ;). The purpose of this joke is to break up the dense technical content with some levity and personal connection to the reader - as such I'd like to keep it in my voice.